cpp/src_prims/metrics/adjustedRandIndex.h

```diff
@@ -0,0 +1,161 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+* @file adjustedRandIndex.h
+* @brief The adjusted Rand index is the corrected-for-chance version of the Rand index.
+* Such a correction for chance establishes a baseline by using the expected similarity 
+* of all pair-wise comparisons between clusterings specified by a random model.
+*/
+
+#include <math.h>
+#include <cub/cub.cuh>
+#include "common/cuml_allocator.hpp"
+#include "common/device_buffer.hpp"
+#include "cuda_utils.h"
+#include "linalg/map_then_reduce.h"
+#include "linalg/reduce.h"
+#include "metrics/contingencyMatrix.h"
+
+namespace MLCommon {
+
+/**
+* @brief Lambda to calculate the number of unordered pairs in a given input
+*
+* @tparam Type: Data type of the input 
+* @tparam IdxType : type of the indexing (by default int)
+* @param in: the input to the functional mapping
+* @param i: the indexing(not used in this case)
+*/
+template <typename Type, typename IdxType = int>
+struct nCTwo {
+  HDI Type operator()(Type in, IdxType i = 0) { return ((in) * (in - 1)) / 2; }
+};
+
+namespace Metrics {
+
+/**
+* @brief Function to calculate Adjusted RandIndex
+* <a href="https://en.wikipedia.org/wiki/Rand_index">more info on rand index</a> 
+* @param firstClusterArray: the array of classes of type T
+* @param secondClusterArray: the array of classes of type T
+* @param size: the size of the data points of type int
+* @param numUniqueClasses: number of Unique classes used for clustering
+* @param lowerLabelRange: the lower bound of the range of labels
+* @param upperLabelRange: the upper bound of the range of labels
+* @param allocator: object that takes care of temporary device memory allocation of type std::shared_ptr<MLCommon::deviceAllocator>
+* @param stream: the cudaStream object
+*/
+template <typename T>
+double computeAdjustedRandIndex(
+  const T* firstClusterArray, const T* secondClusterArray, const int size,
+  const T lowerLabelRange, const T upperLabelRange,
+  std::shared_ptr<MLCommon::deviceAllocator> allocator, cudaStream_t stream) {
+  //rand index for size less than 2 is not defined
+  ASSERT(size >= 2, "Rand Index for size less than 2 not defined!");
+
+  int numUniqueClasses = upperLabelRange - lowerLabelRange + 1;
+
+  //declaring, allocating and initializing memory for the contingency marix
+  MLCommon::device_buffer<int> dContingencyMatrix(
+    allocator, stream, numUniqueClasses * numUniqueClasses);
+  CUDA_CHECK(cudaMemsetAsync(dContingencyMatrix.data(), 0,
+                             numUniqueClasses * numUniqueClasses * sizeof(int),
+                             stream));
+
+  //workspace allocation
+  char* pWorkspace = nullptr;
+  size_t workspaceSz = MLCommon::Metrics::getCMatrixWorkspaceSize(

```

Comment:
Let's use a more descriptive name since this function is sharing the namespace with everything else under metrics.

Reply:
done!